File: app/models/template.rb
class Template
  include Mongoid::Document
  include Mongoid::Timestamps
  include Mongoid::Enum
  
  embeds_many :departments

  field :name, type: String
  field :client_name, type: String
  field :client_address, type: String

  enum :format, [:short,
                 :long,
                 :legacy,
                 :ncmf,
                 :bir,
                 :orlaza], default: :short

  validates_presence_of :name, :client_name
  validates_uniqueness_of :name

  index({ name: 1 }, { unique: true })
end

File: src/app/models/template.ts
import { MongoidId } from './mongoid-id';
import { Timestamps } from './timestamps';
import { Department, DepartmentParams } from './department';

interface Client {
  name: string;
  address?: string;
}

interface TemplatePartials {
  name: string;
  _format?: 'short' | 'long';
  client: Client;
}

export interface TemplateParams extends TemplatePartials {
  departments?: DepartmentParams[];
}

export interface Template extends TemplatePartials, MongoidId, Timestamps {
  departments?: Department[];
}
